File: test/benchmark_xlib.php
<?php
// test/benchmark_xlib.php
// PHP json vs Python json vs lua-yar (cross-language reference)

$sample = ['a' => 1, 'b' => 'hello world', 'c' => [1, 2, 3, 4, 5], 'd' => true, 'e' => 3.14];
$json_str = json_encode($sample);
$msgpack_str = ''; // PHP msgpack not installed

function bench($name, $fn, $n) {
    // warmup
    $warmup = min($n, 1000);
    for ($i = 0; $i < $warmup; $i++) $fn();
    // measure
    $start = hrtime(true);
    for ($i = 0; $i < $n; $i++) $fn();
    $elapsed = (hrtime(true) - $start) / 1e9;
    $ops = $n / $elapsed;
    printf("  %-42s %8d ops in %6.3fs  ->  %10.0f ops/s\n", $name, $n, $elapsed, $ops);
    return $ops;
}

echo "=== PHP json benchmark ===\n";
echo "PHP version: " . PHP_VERSION . "\n";
echo "json extension: " . (function_exists('json_encode') ? 'yes' : 'no') . "\n";
echo "msgpack extension: " . (extension_loaded('msgpack') ? 'yes' : 'no') . "\n";
echo "\n";

echo "[JSON encode]\n";
bench("  php json_encode", function() use ($sample) { json_encode($sample); }, 100000);
echo "\n";

echo "[JSON decode]\n";
bench("  php json_decode", function() use ($json_str) { json_decode($json_str, true); }, 100000);
echo "\n";

if (extension_loaded('msgpack')) {
    $mp_str = msgpack_pack($sample);
    echo "[Msgpack encode]\n";
    bench("  php msgpack_pack", function() use ($sample) { msgpack_pack($sample); }, 100000);
    echo "\n";
    echo "[Msgpack decode]\n";
    bench("  php msgpack_unpack", function() use ($mp_str) { msgpack_unpack($mp_str); }, 100000);
    echo "\n";
}

echo "=== done ===\n";

File: test/benchmark_xlib.py
#!/usr/bin/env python3
# test/benchmark_xlib.py
# Python json benchmark (cross-language reference)

import json
import time

sample = {"a": 1, "b": "hello world", "c": [1, 2, 3, 4, 5], "d": True, "e": 3.14}
json_str = json.dumps(sample)

def bench(name, fn, n):
    # warmup
    for _ in range(min(n, 1000)):
        fn()
    # measure
    start = time.perf_counter()
    for _ in range(n):
        fn()
    elapsed = time.perf_counter() - start
    ops = n / elapsed
    print(f"  {name:42s} {n:8d} ops in {elapsed:6.3f}s  ->  {ops:10.0f} ops/s")
    return ops

print("=== Python json benchmark ===")
print(f"Python version: {__import__('sys').version}")
print()

print("[JSON encode]")
bench("  python json.dumps", lambda: json.dumps(sample), 100000)
print()

print("[JSON decode]")
bench("  python json.loads", lambda: json.loads(json_str), 100000)
print()

try:
    import msgpack
    mp_str = msgpack.packb(sample)
    print("[Msgpack encode]")
    bench("  python msgpack.packb", lambda: msgpack.packb(sample), 100000)
    print()
    print("[Msgpack decode]")
    bench("  python msgpack.unpackb", lambda: msgpack.unpackb(mp_str), 100000)
    print()
except ImportError:
    print("msgpack: NOT INSTALLED")
    print()

print("=== done ===")
